docs/design/sql/06-plan-caching.md

```diff
@@ -0,0 +1,129 @@
+
+The maximum size of the cache is required to prevent out-of-memory if too many distinct queries are submitted.
+
+When a plan is added to the map, the map size is checked. If the map size is greater than the maximum size, some plans are 
+evicted. 
+
+Eviction is synchronous because the asynchronous variant is prone to out-of-memory. We assume that for the most workloads 
+evictions should be rare. 
+
+We use the LRU (least recently used) approach to find the plans to evict. Whenever a plan is accessed, it's `lastUsed` field is 
+updated with the current time. During the eviction, plans are sorted by their `lastUsed` values, and the least recently used
+plans are removed.   
+
+### 2.4 Reacting to Catalog and Partition Distribution Changes
+
+If a catalog or partition distribution is changed, some plans must be invalidated. There are two different ways to achieve 
+this: `push` and `pull`.
+
+With the `push` approach, the plan cache is notified about a change, from the relevant component. E.g., if an index is created,
+then the map service notifies the plan cache about the change. The advantage of this solution is that any change is reflected 
+in the plan cache immediately. However, this approach increases coupling, because many components (map service, 
+replicated map service, partition service) now have to be aware of the SQL subsystem. This approach also requires complex
+synchronization between query optimizer, plan cache, and dependent components, to ensure that no stale plan is ever cached.
+
+With the `pull` approach, the SQL subsystem queries other components periodically, collects the changes, and invalidates 
+affected plans. The advantage of this approach is simplicity. No synchronization or changes to other components are needed. 
+Invalid plans are guaranteed to be removed eventually. The downside is that invalid plans might be active for some time after
+the change has occurred.
+
+We choose the `pull` approach due to simplicity and sufficient guarantees. The background worker reconstructs the catalog 
+periodically, and verifies that existing plans are compatible with the current catalog and partition distribution. 
+
+To counter the problem with outdated plans, we add a special `invalidatePlan` flag to `QueryException`. If an invalid plan
+is used, an exception with this flag will be thrown at some point. When the initiator member receives an exception with 
+this flag, the plan is invalidated. 
+
+Note that currently, users will have to re-execute the query in this case. In future versions, we will add a transparent
+query retry, so that invalid plans will not be visible to users.    
+
+### 2.5 No Use of Parameters
+
+In other databases, parameters are used for plan caching, because the same queries with different parameters may have different 
+optimal plans.
+
+We do not use parameters at the moment, because we do not have statistics. We may change this decision in the future when 
+statistics are available.
```

Comment:
Parameter types are fixed in the plan. They are resolved based on the context. E.g. in this specific example the expected parameter type will be `VARCHAR`. Before the query is executed, the real parameters are converted to expected types.

Reply:
Wanted to test it but found out that parameters are not yet supported. Are they in scope for 4.1? I guess so, they are in the public API.

I wrote this test: 
```java
@Test
public void test() {
    TestHazelcastInstanceFactory factory = new TestHazelcastInstanceFactory();
    HazelcastInstance inst = factory.newHazelcastInstance();
    IMap<Object, Object> map = inst.getMap("map");
    map.put("05", "05");
    map.put("5", "5");

    executeAndPrint(inst, "select __key from map where this=?", "05");
    executeAndPrint(inst, "select __key from map where this=?", 5);
    ((SqlServiceImpl) inst.getSql()).getPlanCache().clear();
    System.out.println("plan cache invalidated");
    executeAndPrint(inst, "select __key from map where this=?", 5);
}

private void executeAndPrint(HazelcastInstance inst, String sql, Object ... params) {
    System.out.println("params: " + Arrays.stream(params)
                                          .map(p -> p.getClass().getSimpleName() + " " + p)
                                          .collect(Collectors.joining(",")));
    for (SqlRow r : inst.getSql().query(sql, params)) {
        System.out.println(r);
    }
    System.out.println("----");
}
```
I expect this output:
```
params: String 05
[__key VARCHAR=05]
params: Integer 5
[__key VARCHAR=5]
plan cache invalidated
params: Integer 5
[__key VARCHAR=05]
[__key VARCHAR=5]
```
That is, the first query will have VARCHAR parameter and will get cached, including the parameter type. The second query will have INT parameter. The cached plan will be used, the INT will be casted to VARCHAR and used in the filter. After invalidating the plan cache, new plan will be created, this time with INT parameter. The `field` will be casted to INT and this time, both entries will be returned.

Reply:
As explained above, parameter values do not participate in plan creation. Whatever parameter values are passed, the created plan will be the same.

Reply:
I tried it in the expressions PR and indeed it seems that the parameter is converted to a type based on the expression it appears in. This is really surprising to me. The consequence is that implicit conversion rules of SQL do not apply as in my example above: if you have a VARCHAR field and compare it to NUMBER parameter, the number parameter gets converted to VARCHAR and not the other way as would be the case if the number was a literal.

Then I tried `SELECT ? FROM map` and got `Illegal use of dynamic parameter` - obviously. Also `SELECT a FROM t WHERE f=? + ?` throws `Cannot apply '+' to arguments of type '<UNKNOWN> + <UNKNOWN>`.

If this is the behavior of expressions, then this PR is correct, we don't need to store parameter types in plan cache key. But I think this behavior is not correct. Checked with PostgreSQL, parameter type is taken into account. You also can do `SELECT ?`. I don't have other DB at hand locally, but [this fiddle](https://dbfiddle.uk/?rdbms=oracle_18&fiddle=f2dd3168e778ad97c06a707e745c48f7) with oracle suggests that oracle also takes the type of dynamic parameter into account: queries in PL/SQL blocks are converted to ones with parameters.

Also it is not as JDBC is designed. In JDBC you set the parameters in a typed way, using `PreparedStatement.setInt()`, `setString()` etc. I checked a few JDBC implementations, all but HSQLDB store the type used to set the parameter. Mysql even replaces the parameter with a literal.

cc: @taburet

Reply:
Parameters are not used in planning for a reason - there will be too many different plans otherwise. We would like to avoid this complexity unless there is an obvious reason to go into it. There is no correct or wrong way of handling parameters wrt the ANSI standard. We are free to choose the way we find it convenient.

Isolated usage of parameters (e.g. `SELECT ? ...`, or `? + ?`) will work in future releases, there is an issue for this. Things like `SELECT varchar + ?` attempts to use the existing context to derive the result, the correct semantics is yet to be understood during beta, but, for example, assuming that the parameter is DECIMAL would be reasonable for the most cases. And the main thing - if the user is not satisfied with the result, he can always use `CAST` operator.

Regarding JDBC - this is not really the case. This is the user who defines parameter types. Parameter types are defined by the database before the actual values are provided. The database gives users information about expected parameter types through `PreparedStatement.getParameterMetadata`. Nevertheless, this doesn't mean that result set metadata is fixed irrespectively of actual parameters. For example, some databases (e.g. Postgres) may decide to use a special `UNKNOWN` type for parameter placeholder for the statement `SELECT ?`, and the result set metadata will be different for different parameter values. Other vendors may use fixed types assigned to parameters. None of these approaches violate ANSI SQL or JDBC to my knowledge.

Reply:
>Parameters are not used in planning for a reason - there will be too many different plans otherwise

We can cache the plan based just on the parameter type, not value. Then there won't be that many plans for each SQL string, one most of the time.

While it's probably true that neither ANSI nor JDBC specify this detail, I still think we should consider using the type from the client in planning. As you see, the current approach opens many questions. It's the question whether a parameter is a drop-in replacement for a literal or not.

Reply:
I'm marking this as resolved as it's not related to this PR but rather to #17183